src/backend_server.h

```diff
@@ -0,0 +1,39 @@
+#ifndef CHIRP_SRC_BACKEND_SERVER_H_
+#define CHIRP_SRC_BACKEND_SERVER_H_
+
+#include <atomic>
+#include <map>
+#include <string>
+
+#include <grpc/grpc.h>
+#include <grpcpp/server.h>
+#include <grpcpp/server_builder.h>
+#include <grpcpp/server_context.h>
+#include <grpcpp/security/server_credentials.h>
+
+#include "backend_data_structure.h"
+#include "key_value.grpc.pb.h"
+
+// Key-value store implementation inherits from the `chirp::KeyValueStore::Service`
+// which implements the `put`, `get`, and `deletekey` operations
+class KeyValueStoreImpl final : public chirp::KeyValueStore::Service {
+ public:
+  explicit KeyValueStoreImpl();
+
+  // Accepts put requests
+  grpc::Status put(grpc::ServerContext *context, const chirp::PutRequest *request, chirp::PutReply *reply) override;
+
+  // Accepts get requests
+  grpc::Status get(grpc::ServerContext *context, grpc::ServerReaderWriter<chirp::GetReply, chirp::GetRequest> *stream) override;
+
+  // Accepts deletekey requests
+  grpc::Status deletekey(grpc::ServerContext *context, const chirp::DeleteRequest *request, chirp::DeleteReply *reply) override;
+
+ private:
+  BackendDataStructure backend_data_;
+
+  // spinlock
+  std::atomic_flag lock_;
```

Comment:
would it make more sense to have the KV store own the lock? I feel that the server is responsible for the communication between the service layer while the actual backend_data_structure should be responsible for handling multiple requests

Reply:
This makes sense.

Reply:
As I tried to fix this, I found a problem. In my KV store implementation, a get means getting a value from the KV store. But in the server, a get means getting multiple values from the KV store behind. Therefore, I think the critical section should depend on every request from a grpc client. Currently, I am not going to change this.